### PR TITLE
bpo-33191: Fix refleak in posix_spawn

### PR DESCRIPTION
**Pull request by zware**
https://bugs.python.org/issue33191

### CODE REVIEW

File: Modules/posixmodule.c [hunk 1]
```diff
@@ -5194,17 +5195,17 @@ os_posix_spawn_impl(PyObject *module, path_t *path, PyObject *argv,
...
         }
-        PyObject* file_actions_obj;
         PyObject* mode_obj;
 
         for (int i = 0; i < PySequence_Fast_GET_SIZE(seq); ++i) {
             file_actions_obj = PySequence_Fast_GET_ITEM(seq, i);
-
-            if(!PySequence_Check(file_actions_obj) | !PySequence_Size(file_actions_obj)) {
-                PyErr_SetString(PyExc_TypeError,"Each file_action element must be a non empty sequence");
+            if (file_actions_obj == NULL) {
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:5202]: Is it possible?

File: Modules/posixmodule.c [hunk 2]
```diff
@@ -5194,17 +5195,17 @@ os_posix_spawn_impl(PyObject *module, path_t *path, PyObject *argv,
...
+                goto exit;
+            }
+            file_actions_obj = PySequence_Fast(file_actions_obj, "Each file_action element must be a non-empty sequence");
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:5205]: Would be nice to wrap this long line.

File: Modules/posixmodule.c [hunk 3]
```diff
@@ -5219,19 +5220,19 @@ os_posix_spawn_impl(PyObject *module, path_t *path, PyObject *argv,
                         goto exit;
                     }
 
-                    long open_fd = PyLong_AsLong(PySequence_GetItem(file_actions_obj, 1));
+                    long open_fd = PyLong_AsLong(PySequence_Fast_GET_ITEM(file_actions_obj, 1));
                     if(PyErr_Occurred()) {
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:5224]: While you are here, could you please make the source conforming PEP 7: Add spaces after "if" and "switch", wrap long lines?

File: Modules/posixmodule.c [hunk 4]
```diff
@@ -5219,19 +5220,19 @@ os_posix_spawn_impl(PyObject *module, path_t *path, PyObject *argv,
                         goto exit;
                     }
 
-                    long open_fd = PyLong_AsLong(PySequence_GetItem(file_actions_obj, 1));
+                    long open_fd = PyLong_AsLong(PySequence_Fast_GET_ITEM(file_actions_obj, 1));
```
Comment (serhiy-storchaka) [inline Modules/posixmodule.c:5223]: `PyLong_AsLong()` can call arbitrary code. If `file_actions` is a list, the size of `file_actions_obj` can be changed after this, and the following `PySequence_Fast_GET_ITEM()` could read past the end of the list.

I would require `file_actions` to be a tuple and use `PyArg_ParseTuple()` for parsing it.

**Comment by serhiy-storchaka** [comment]
I have added more comments in #5109 because it is hard to add comment to the code not changed in a PR.

**Comment by zware** [comment]
@serhiy-storchaka Would you mind to take this over? I would like to see this fixed before 3.7.0 final, but don't know when I'll be able to get back to it.

**Comment by pablogsal** [comment]
I have opened #6331 trying to address al the issues at once.

**Comment by gpshead** [comment]
https://github.com/python/cpython/pull/6332 should obsolete this

**Comment by zware** [comment]
Superseded by #6332 and/or #6331.

**Pull request closed without merging**